src/io/RtInputScannerImages.cpp

```diff
@@ -451,6 +522,135 @@ short *RtInputScannerImages::receiveImage(ACE_SOCK_Stream &stream,
   return img;
 };
 
+// read the next unread image from disk
+RtMRIImage* RtInputScannerImages::readImageFromDICOMFolder() {
+  static set<string> alreadyRead;
+
+  // first time we are called, mark all previously existing files as read
+  static bool firstCall = true;
+  if(firstCall) {
+    for(const auto &entry: std::filesystem::recursive_directory_iterator(dicomDir)) {
+      if(entry.is_directory()) {
+        continue;
+      }
+
+      alreadyRead.insert(entry.path().string());
+    }
+    firstCall = false;
+    return NULL;
+  }
+
+  // created a sorted set of the files in the dir
+  set<filesystem::path> files;
+  for(const auto &entry: filesystem::recursive_directory_iterator(dicomDir)) {
+    if(entry.is_directory()) {
+      continue;
+    }
+
+    files.insert(entry.path());
+  }
+
+  // get the set difference between the files and the already read files
+  set<filesystem::path> notYetRead;
+  set_difference(files.begin(), files.end(), alreadyRead.begin(), alreadyRead.end(), inserter(notYetRead, notYetRead.begin()));
+
+  // if there are no files to read, return NULL
+  if(notYetRead.empty()) {
+    return NULL;
+  }
+
+  // get the first file to read
+  string toRead = notYetRead.begin()->string();
+  cout << "found next dicom file to read as input: " << toRead << endl;
+
+  // even though we may fail to read the image, we don't want to try again
+  // because we have no reason to think it will succeed in the future
+  alreadyRead.insert(toRead);
+
+  // try to read the image multiple times. this is a hack to deal with the
+  // fact that dcm2niix sometimes isn't finished writing the image before we try
+  // to read it
+  stringstream ss;
+  for(int tryNum = 0; tryNum < MAX_DCM2NIIX_TRIES; tryNum++) {
+    // invoke dcm2niix to convert the single image in verbose mode to /tmp
+    string cmd = "dcm2niix -v y -f '%f_%p_%t_%s_%u' -s y -o /tmp " + toRead;
+    cout << "(try number " << tryNum + 1 << " / 10), executing: " << cmd << endl;
+
+    FILE *pipe = popen(cmd.c_str(), "r");
+    if(!pipe) {
+      cerr << "failed to create pipe." << endl;
```

Comment:
Can we output the reason the pipe creation fails?  Would be good to better understand what's going on here.

```
    cerr << "Failed to create pipe: " << strerror(errno) 
         << " (errno: " << errno << ")" << endl;
```

Reply:
done